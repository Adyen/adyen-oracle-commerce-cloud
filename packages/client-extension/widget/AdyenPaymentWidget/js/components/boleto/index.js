import * as constants from '../../constants'
import {
    addDays,
    Checkout,
    createPresentToShopperModal,
    eventEmitter,
    createFromAction,
} from '../../utils'
import { store } from '../index'

export const setBoletoConfig = ({
    paymentMethodTypes,
    boletoDeliveryDate,
    boletoShopperStatement,
}) => {
    const hasBoleto = paymentMethodTypes.includes(
        constants.paymentMethodTypes.invoice
    )
    eventEmitter.store.emit(constants.boleto.enabled, hasBoleto)

    if (hasBoleto) {
        boletoDeliveryDate &&
            eventEmitter.store.emit(
                constants.boleto.deliveryDate,
                boletoDeliveryDate
            )
        boletoShopperStatement &&
            eventEmitter.store.emit(
                constants.boleto.shopperStatement,
                boletoShopperStatement
            )
    }
}

export const presentToShopper = customPaymentProperties => {
    const checkoutBoleto = store.get(constants.checkout.boleto)
    const options = {
        action: customPaymentProperties,
        selector: '#present-shopper',
        checkoutComponent: checkoutBoleto,
    }

    createPresentToShopperModal(() => {
        createFromAction(options)
    })
}

const createBoletoCheckout = () => {
    const checkout = new Checkout(constants.paymentMethodTypes.invoice)
    const onChange = checkout.onChange({
        deliveryDate: addDays(store.get(constants.boleto.deliveryDate)),
        shopperStatement: store.get(constants.boleto.shopperStatement),
    })

    const onSubmit = checkout.onSubmit()

    const configuration = { onSubmit, onChange }
    const order = store.get(constants.order)

    const address = order().billingAddress()

    const shopperName = {
        firstName: address.firstName(),
        lastName: address.lastName(),
    }
    const billingAddress = {
        city: address.city(),
        country: address.selectedCountry(),
        // houseNumberOrName:"952",
        postalCode: address.postalCode(),
        stateOrProvince: address.selectedState(),
        street: address.address1(),
    }
    const options = { data: { shopperName, billingAddress } }
    const checkoutOptions = {
        configuration,
        selector: '#adyen-boleto-payment',
        type: 'boletobancario',
        options,
    }

    checkout.createCheckout(checkoutOptions, checkout => {
        eventEmitter.store.emit(constants.checkout.boleto, checkout)
    })
}

export default createBoletoCheckout
