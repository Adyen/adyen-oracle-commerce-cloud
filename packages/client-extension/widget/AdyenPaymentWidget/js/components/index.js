export { default as Component } from './component'
export { default as Order } from './order'
export { default as Payment } from './payment'
export { default as store } from './store'
export {
    default as createBoletoCheckout,
    presentToShopper,
    setBoletoConfig,
} from './boleto'
export { default as createCardCheckout } from './card'
