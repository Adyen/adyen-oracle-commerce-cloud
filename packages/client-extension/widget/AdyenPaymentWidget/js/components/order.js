import $ from 'jquery'
import ccRestClient from 'ccRestClient'
import ccConstants from 'ccConstants'
import storageApi from 'storageApi'
import pubsub from 'pubsub'
import {
    redirectAuth,
    createSpinner,
    getOrderPayload,
    eventEmitter,
} from '../utils'
import * as constants from '../constants'
import { store } from './index'

class Order {
    constructor() {
        this.instance = storageApi.getInstance()

        this.startEventListeners()
    }

    startEventListeners = () => {
        eventEmitter.order.on(
            constants.pageChanged,
            this.getUrlParametersAndCreateOrder
        )
        eventEmitter.order.on(
            constants.initialOrderCreated,
            this.initialOrderCreated
        )
    }

    initialOrderCreated = orderEvent => {
        const order = store.get(constants.order)
        const {
            customPaymentProperties: { data, resultCode },
        } = orderEvent.order.payments[0]

        const payload = {
            order: getOrderPayload(order),
            customPaymentProperties: JSON.parse(data),
            resultCode,
        }

        eventEmitter.store.emit(constants.orderPayload, payload)
        const isDone = store.get(constants.isDone)
        !isDone() && redirectAuth(payload, this.createOrder)
    }

    getAndCreatOrder = result => {
        result.paymentData = this.instance.getItem(
            constants.storage.paymentData
        )

        const payment = { type: 'generic', customProperties: result }

        this.recreateOrder(payment)
    }

    createOrder = orderPayload => {
        const translate = store.get(constants.translate)
        const orderFail = ({ message = 'Failed to create order' } = {}) => {
            $.Topic(pubsub.topicNames.ORDER_SUBMISSION_FAIL).publish({
                message: 'fail',
                errorMessage: message,
            })
        }

        const submitOrder = ({ id, uuid }) => {
            const publishData = { message: 'success', id, uuid }
            $.Topic(pubsub.topicNames.ORDER_COMPLETED).publish(publishData)
            $.Topic(pubsub.topicNames.ORDER_SUBMISSION_SUCCESS).publish([
                publishData,
            ])
        }

        const orderSuccess = data => {
            const { paymentState } = data.payments[0]
            const isAuthorized =
                paymentState === ccConstants.PAYMENT_GROUP_STATE_AUTHORIZED
            const message = translate(
                constants.errorMessages.failed3dsValidation
            )
            isAuthorized ? submitOrder(data) : orderFail({ message })
        }

        ccRestClient.request(
            ccConstants.ENDPOINT_ORDERS_CREATE_ORDER,
            orderPayload,
            orderSuccess,
            orderFail
        )
    }

    recreateOrder = paymentData => {
        const storedOrder = JSON.parse(
            this.instance.getItem(constants.storage.order)
        )
        storedOrder.payments = [paymentData]

        eventEmitter.store.emit(constants.isDone, true)
        this.createOrder(storedOrder)
    }

    getResult = urlParameters => {
        const params = urlParameters.split('&')
        const formatResult = (acc, param) => {
            const [key, value] = param.split('=')
            return { ...acc, [key]: decodeURIComponent(value) }
        }

        return params.reduce(formatResult, {})
    }

    getUrlParametersAndCreateOrder = pageData => {
        const order = store.get(constants.order)
        const { parameters } = pageData

        if (parameters) {
            createSpinner()

            const parametersObj = this.getResult(parameters)

            if (order().paymentGateway()) {
                order().paymentGateway().type = ''
            }

            this.isAction = 'PaRes' in parametersObj && 'MD' in parametersObj
            this.isAction && this.getAndCreatOrder(parametersObj)
        }
    }
}

export default Order
