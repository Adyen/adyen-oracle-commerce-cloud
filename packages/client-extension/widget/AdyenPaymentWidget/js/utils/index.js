export { createSpinner, destroySpinner } from './spinner'
export { default as ajax } from './ajax'
export { default as redirectAuth } from './redirectAuth'
export { default as getOrderPayload } from './getOrderPayload'
export { default as eventEmitter } from './events'
export { default as getInstallmentOptions } from './getInstallmentOptions'
export { default as Checkout, createFromAction } from './checkout'
export { createPresentToShopperModal } from './modal'
export { default as createError } from './createError'
export { addDays } from './date'
