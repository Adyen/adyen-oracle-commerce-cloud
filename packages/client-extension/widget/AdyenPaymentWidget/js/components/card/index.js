import { store } from '../index'
import * as constants from '../../constants'
import { Checkout, eventEmitter, getInstallmentOptions } from '../../utils'

export const onBrand = ({ brand }) => {
    eventEmitter.store.emit(constants.selectedBrand, brand)
    const installmentsOptions = store.get(constants.installmentsOptions)
    const cart = store.get(constants.cart)
    const amount = cart().amount()
    const installments = installmentsOptions ? getInstallmentOptions(installmentsOptions, amount, brand) : []

    eventEmitter.store.emit(constants.installments, installments)
}

export const onConfigSuccess = () => eventEmitter.store.emit(constants.isLoaded, true)

const createCardCheckout = () => {
    // const [card] = paymentMethods.filter(paymentMethod => paymentMethod.type === constants.paymentMethodTypes.scheme)

    // if (card) {
    //     eventEmitter.store.emit(constants.paymentMethodTypes.scheme, true)
    //     const { brands } = card
    const checkout = new Checkout(constants.paymentMethodTypes.scheme)
    const onSubmit = checkout.createOnSubmit()
    const onChange = checkout.createOnChange()
    const cardComponent = document.querySelector('adyen-payment-method-card')
    cardComponent.addEventListener('adyenChange', onChange)
    cardComponent.addEventListener('adyenSubmit', onSubmit)

    eventEmitter.store.emit(constants.checkout.card, checkout)
    // const cardComponent = document.querySelector('adyen-payment-method-card')

    //     const onChange = checkout.onChange()
    //     const onSubmit = checkout.onSubmit()

        // const storedPaymentType = store.get(constants.storedPaymentType)
        // const options = { hasHolderName: false, enableStoreDetails: !!storedPaymentType(), brands }

        // const configuration = { onBrand, onConfigSuccess, onChange, onSubmit }
        // const checkoutOptions = { configuration, selector: '#adyen-card-payment', type: 'card', options }

        // checkout.createCheckout(checkoutOptions, checkout => {
        //     eventEmitter.store.emit(constants.checkout.card, checkout)
        // })
    // }
}

export default createCardCheckout
