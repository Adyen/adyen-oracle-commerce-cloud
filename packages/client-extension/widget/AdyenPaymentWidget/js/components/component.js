import * as constants from '../constants'
import { eventEmitter } from '../utils'
import { createBoletoCheckout, createCardCheckout, store } from './index'

class Component {
    constructor() {
        this.startEventListeners()
    }

    startEventListeners = () => {
        eventEmitter.component.on(constants.render, this.render)
        eventEmitter.component.on(
            constants.comboCardOptions,
            this.getComboCardOptions
        )
    }

    getComboCardOptions = () => {
        const brands = { visa: constants.bins.electron }
        const selectedComboCard = store.get(constants.selectedComboCard)()
        const isDebitCard = selectedComboCard === constants.comboCards.debit
        const brand = store.get(constants.selectedBrand)
        const selectedBrand =
            brand in brands
                ? brands[store.get(constants.selectedBrand)]
                : store.get(constants.selectedBrand)

        const options = {
            selectedBrand,
            additionalData: { overwriteBrand: true },
        }

        eventEmitter.store.emit(
            constants.comboCardOptions,
            isDebitCard ? options : {}
        )
    }

    render = () => {
        store.get(constants.ajax)('paymentMethods', this.getPaymentMethods)
    }

    getPaymentMethods = paymentMethodsResponse => {
        eventEmitter.store.emit(
            constants.paymentMethodsResponse,
            paymentMethodsResponse
        )

        createCardCheckout()
        store.get(constants.brazilEnabled) && createBoletoCheckout()
    }
}

export default Component
