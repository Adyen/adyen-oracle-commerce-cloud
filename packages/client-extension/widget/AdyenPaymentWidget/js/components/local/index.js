import * as constants from '../../constants'
import { Checkout, eventEmitter } from '../../utils'

const hasType = paymentMethod => 'type' in paymentMethod
const isNotScheme = paymentMethod => hasType(paymentMethod) && paymentMethod.type !== 'scheme'
const hasDetails = paymentMethod => isNotScheme(paymentMethod) && 'details' in paymentMethod
const hasOneItem = paymentMethod => hasDetails(paymentMethod) && paymentMethod.details.length === 1
const getDetail = paymentMethod => hasOneItem(paymentMethod) && paymentMethod.details[0]
const isValid = (paymentMethod, key, value) => getDetail(paymentMethod) && getDetail(paymentMethod)[key] === value

const isLocalPaymentMethod = paymentMethod => {
    const isIssuer = isValid(paymentMethod, 'key', 'issuer')
    const isSelect = isValid(paymentMethod, 'type', 'select')
    const isLocal = isIssuer && isSelect

    return isLocal
}

const createLocalComponent = ({ type, name }) => {
    const headingNode = document.createElement('div')
    headingNode.classList.add('collapse-heading')
    headingNode.innerHTML = ` <strong>
        <span
            class="collapse-toggle collapsed"
            data-toggle="collapse"
            data-target="#adyen-${type}-wrapper"
        >
          <button>${name}</button>
        </span>
      </strong>`
    const component = document.createElement('div')
    component.setAttribute('id', `adyen-${type}-wrapper`)
    component.classList.add('adyen-wrapper', 'collapse')

    component.innerHTML = `
        <div class="loader-wrapper hide">
            <div class="lds-ring"><div></div><div></div><div></div><div></div></div>
            <div style="font-weight: bold; font-size: 16px;">Processing...</div>
        </div>
        <div id="adyen-${type}-payment" class="adyen-component"></div>
    `
    const adyenNode = document.getElementById('adyen-collapse-heading')
    adyenNode.appendChild(headingNode)
    adyenNode.appendChild(component)
}

const createLocalCheckout = (acc, localPaymentMethod) => {
    const { type } = localPaymentMethod
    const checkout = new Checkout(type)
    const onChange = checkout.onChange()
    const onSubmit = checkout.onSubmit()

    const configuration = { onSubmit, onChange, showPayButton: true }

    createLocalComponent(localPaymentMethod)
    const checkoutOptions = type => ({ configuration, selector: `#adyen-${type}-payment`, type })
    checkout.createCheckout(checkoutOptions(type), checkout => ({ ...acc, [type]: checkout }))
}

const createLocalPaymentCheckout = ({ paymentMethods }) => {
    const localPaymentMethods = paymentMethods.filter(isLocalPaymentMethod)
    const localPaymentMethodsCheckout = localPaymentMethods.reduce(createLocalCheckout, {})

    eventEmitter.store.emit(constants.checkout.local, localPaymentMethodsCheckout)
}

export default createLocalPaymentCheckout
