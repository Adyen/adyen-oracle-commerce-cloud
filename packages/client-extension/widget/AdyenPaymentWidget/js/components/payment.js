import * as constants from '../constants'
import { eventEmitter } from '../utils'
import { store } from './index'

class Payment {
    constructor() {
        this.startEventListeners()
    }

    startEventListeners = () => {
        eventEmitter.payment.on(constants.setPayment, this.setPayment)
    }

    setPayment = type => {
        eventEmitter.store.emit(constants.isSubmitting, true)
        const order = store.get(constants.order)
        const id = store.get(constants.id)

        const genericPayment = store.get(constants.genericPayment)
        const updatedGenericPayment = { ...genericPayment, id: id() }

        eventEmitter.store.emit(constants.genericPayment, updatedGenericPayment)
        eventEmitter.component.emit(constants.comboCardOptions)

        const payments = [updatedGenericPayment]

        order().updatePayments(payments)

        this.updatePayment(this.createPayment(type))
    }

    updatePayment = pPayment => {
        const order = store.get(constants.order)
        const id = store.get(constants.id)

        const payments = order().payments()
        const setPayments = (payment, index) => {
            const curPayment = payments[index]
            const { id: curId } = curPayment
            const isSameId = curId && curId === id()
            return isSameId ? pPayment : curPayment
        }

        const newPayments = payments.map(setPayments)
        order().payments(newPayments)
    }

    createPayment = type => {
        const storedPaymentType = store.get(constants.storedPaymentType)
        const isPaymentStored = store.get(constants.isPaymentStored)
        const paymentDetails = store.get(constants.paymentDetails)
        const selectedInstallment = store.get(constants.selectedInstallment)
        const selectedComboCard = store.get(constants.selectedComboCard)
        const comboCardOptions = store.get(constants.comboCardOptions)

        const hasInstallments =
            selectedInstallment() &&
            'numberOfInstallments' in selectedInstallment()

        const isCreditCard = selectedComboCard() === constants.comboCards.credit
        const isValidInstallment = hasInstallments && isCreditCard
        const getInstallments = () =>
            isValidInstallment &&
            parseInt(selectedInstallment().numberOfInstallments)

        const storedPayment = isPaymentStored ? storedPaymentType() : ''
        const numberOfInstallments = getInstallments()

        const genericPayment = store.get(constants.genericPayment)
        const updatedGenericPayment = {
            ...genericPayment,
            customProperties: {
                paymentDetails: paymentDetails[type],
                storedPayment,
                ...(numberOfInstallments && { numberOfInstallments }),
                ...comboCardOptions,
            },
        }

        return updatedGenericPayment
    }
}

export default Payment
