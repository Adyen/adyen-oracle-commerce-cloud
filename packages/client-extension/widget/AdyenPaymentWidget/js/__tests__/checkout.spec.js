import { store } from '../components'
import * as constants from '../constants'
import Widget from '../../../../__mocks__/widget'
import ccConstants from '../../../../__mocks__/ccConstants'
import viewModel from '../adyen-checkout'
import Checkout, { createFromAction, getDefaultConfig } from '../utils/checkout'
import { eventEmitter } from '../utils'

const defaultConfig = {
    environment: 'test',
    locale: 'en_US',
    originKey: '',
    paymentMethodsResponse: undefined,
    showPayButton: true,
}

describe('Checkout', () => {
    let widget
    beforeEach(() => {
        widget = new Widget()
    })

    it('should return default config', function() {
        viewModel.onLoad(widget)
        const result = getDefaultConfig()
        expect(result).toEqual(defaultConfig)
    })

    it('should create from action', function() {
        viewModel.onLoad(widget)
        const mount = jest.fn()
        const createFromActionMock = jest.fn(() => ({ mount }))
        const checkoutComponent = { createFromAction: createFromActionMock }
        const args = {
            action: 'mocked_action',
            selector: '#id',
            checkoutComponent,
        }
        createFromAction(args)
        expect(createFromActionMock).toHaveBeenCalledWith(args.action)
        expect(mount).toHaveBeenCalledWith(args.selector)
    })

    it('should create checkout', function() {
        const defaultConfig = getDefaultConfig()
        const configuration = { data: 'mocked_extra_config' }
        const cb = jest.fn()
        const mount = jest.fn()
        const create = jest.fn(() => ({ mount }))
        global.AdyenCheckout = jest.fn(() => ({ create }))
        const checkout = new Checkout(constants.paymentMethodTypes.generic)
        const selector = '#id'
        const type = 'boleto'
        checkout.createCheckout({ configuration, type, selector }, cb)

        expect(create).toHaveBeenCalledWith(type, {})
        expect(mount).toHaveBeenCalledWith(selector)
        expect(global.AdyenCheckout).toHaveBeenCalledWith({
            ...defaultConfig,
            ...configuration,
        })
        expect(cb).toHaveBeenCalled()
    })

    it('should handle on submit', function() {
        eventEmitter.store.emit(constants.paymentDetails, {})
        const checkout = new Checkout(constants.paymentMethodTypes.generic)
        const createOnSubmit = checkout.onSubmit()
        createOnSubmit()

        const order = store.get(constants.order)
        expect(order().id()).toEqual(null)
        expect(order().op()).toEqual(ccConstants.ORDER_OP_INITIATE)
        expect(order().handlePlaceOrder).toHaveBeenCalled()
    })

    it('should handle on change', function() {
        eventEmitter.store.emit(constants.paymentDetails, {})
        const checkout = new Checkout(constants.paymentMethodTypes.generic)
        const createOnSubmit = checkout.onChange()
        const data = { foo: 'bar' }
        createOnSubmit({ data }, { isValid: true })

        const paymentDetails = store.get(constants.paymentDetails)
        const isValid = store.get(constants.isValid)
        const expected = { [constants.paymentMethodTypes.generic]: data }

        expect(isValid).toBeTruthy()
        expect(paymentDetails).toEqual(expected)
    })
})
