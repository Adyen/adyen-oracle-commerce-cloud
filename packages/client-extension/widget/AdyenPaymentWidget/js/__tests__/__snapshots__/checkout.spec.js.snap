// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkout should create checkout 1`] = `
Array [
  Array [
    Object {
      "amount": Object {
        "currency": "USD",
        "value": 1000,
      },
      "countryCode": "US",
      "environment": "test",
      "originKey": "",
      "paymentMethodsResponse": undefined,
      "showPayButton": true,
    },
  ],
]
`;

exports[`Checkout should create stored payments checkout 1`] = `
Array [
  Array [
    Object {
      "amount": Object {
        "currency": "USD",
        "value": 1000,
      },
      "countryCode": "US",
      "environment": "test",
      "originKey": "",
      "paymentMethodsResponse": undefined,
      "showPayButton": true,
    },
  ],
]
`;

exports[`Checkout should return default config 1`] = `
Object {
  "amount": Object {
    "currency": "USD",
    "value": 1000,
  },
  "countryCode": "US",
  "environment": "test",
  "originKey": "",
  "paymentMethodsResponse": undefined,
  "showPayButton": true,
}
`;
